source/daplink/interface/target_flash.c

```diff
@@ -89,20 +115,41 @@ static error_t flash_func_start(flash_func_t func)
     return ERROR_SUCCESS;
 }
 
+static error_t target_flash_set(uint32_t addr)
+{
+    program_target_t * new_flash_algo = get_flash_algo(addr);
+    if(last_flash_algo != new_flash_algo){
+        //run uninit to last func
+        error_t status = flash_func_start(FLASH_FUNC_NOP);
+        if (status != ERROR_SUCCESS) {
+            return status;
+        }
+        // Download flash programming algorithm to target
+        if (0 == swd_write_memory(new_flash_algo->algo_start, (uint8_t *)new_flash_algo->algo_blob, new_flash_algo->algo_size)) {
+            return ERROR_ALGO_DL;
+        }
+        
+        last_flash_algo = new_flash_algo;
+        
+    }
+    return ERROR_SUCCESS;
+}
+
 static error_t target_flash_init()
 {
     if (g_board_info.target_cfg) {
-        const program_target_t *const flash = g_board_info.target_cfg->flash_algo;
-        
         last_flash_func = FLASH_FUNC_NOP;
         
+        last_flash_algo = NULL;
+        
         if (0 == target_set_state(RESET_PROGRAM)) {
             return ERROR_RESET;
         }
 
-        // Download flash programming algorithm to target and initialise.
-        if (0 == swd_write_memory(flash->algo_start, (uint8_t *)flash->algo_blob, flash->algo_size)) {
-            return ERROR_ALGO_DL;
+        //try to set the initial flash algo
+        error_t status = target_flash_set(g_board_info.target_cfg->flash_start); 
```

Comment:
Hex programming can jump to any address, this just sets the default flash algo, in case there are multiple flash regions using the default flash algo or g_board_info.target_cfg->flash_algo